Intro to build servers tali (#613)
OctopusDeploy/blog: blog/2022-01/an-introduction-to-build-servers/index.md

---

> title: An introduction to build servers and continuous integration

Author: ```suggestion
title: An introduction to build servers and Continuous Integration
```
```suggestion
title: An introduction to build servers and continuous integration
```

isFeatured: false
tags:
  - DevOps
  - CI Series
  - Continuous Integration
  - Jenkins
  - GitHub Actions
---

When you're developing and deploying software, one of the first things to figure out is how to take your code and deploy your working application to a production environment where people can interact with your software.

Most development teams understand the importance of version control to coordinate code commits, and build servers to compile and package their software, but Continuous Integration (CI) is a big topic. Over the next few months, we’re going into detail about Continuous Integration and how two of the most popular build servers, Jenkins and GitHub Actions, can help with your CI processes.

## Why build servers are important

Build servers have 3 main purposes:

- Compiling committed code from your repository many times a day
- Running automatic tests to validate code
- Creating deployable packages and handing off to your deployment tool

Without a build server you're slowed down by complicated, manual processes and the needless time constraints they introduce. For example, without a build server:

> - Your team will likely need to commit code before a daily deadline or during change windows.

Author: ```suggestion
- Your team will likely need to commit code before a daily deadline or during change windows
```

> - Once that deadline passes, no one can commit again until someone manually creates and tests a build.

Author: ```suggestion
- After that deadline passes, no one can commit again until someone manually creates and tests a build
```

> - If there are problems with code, the deadlines and manual processes delay the fixes further.

Author: ```suggestion
- If there are problems with code, the deadlines and manual processes further delay the fixes
```

---

> title: An introduction to build servers

Reviewer: ```suggestion
title: An introduction to build servers and continuous integration
```

isFeatured: false
tags:
  - DevOps

>   - CI series

Author: ```suggestion
  - CI Series
```

  - Continuous Integration
  - Jenkins
  - GitHub Actions
---

> Octopus can receive packages whether you upload them to our in-built package store or import them via a feed. As believers in Continuous Integration and Continuous Delivery (CI/CD), though, we think packages should come from a build server, also known as a Continuous Integration (CI) platform.

Reviewer: ```suggestion
When you're developing and deploying software, one of the first things to figure out is how to take your code and deploy your working application to a production environment where people can interact with your software.

Most development teams understand the importance of version control to coordinate code commits, and build servers to compile and package their software, but Continuous Integration (CI) is a big topic. Over the next few months, we’re going into detail about Continuous Integration and how two of the most popular build servers, Jenkins and GitHub Actions, can help with your CI processes.
```

isFeatured: false
tags:
  - DevOps
  - CI Series
  - Continuous Integration
  - Jenkins
  - GitHub Actions
---

When you're developing and deploying software, one of the first things to figure out is how to take your code and deploy your working application to a production environment where people can interact with your software.

Most development teams understand the importance of version control to coordinate code commits, and build servers to compile and package their software, but Continuous Integration (CI) is a big topic. Over the next few months, we’re going into detail about Continuous Integration and how two of the most popular build servers, Jenkins and GitHub Actions, can help with your CI processes.

> Given build servers can automate everything that happens between code commits and deployments, they're vital to CI as a concept. Let’s dig into why we rate them so highly.

Reviewer: ```suggestion

```

  - Continuous Integration
  - Jenkins
  - GitHub Actions
---

Octopus can receive packages whether you upload them to our in-built package store or import them via a feed. As believers in Continuous Integration and Continuous Delivery (CI/CD), though, we think packages should come from a build server, also known as a Continuous Integration (CI) platform.

> 
> Given build servers can automate everything that happens between code commits and deployments, they're vital to CI as a concept. Let’s dig into why we rate them so highly.

Reviewer: Let's add a bit more context here. The piece jumps straight into what octopus can consume, which is important, but as we're targeting non octopus users, we can leave that bit for later :)  For instance:

```suggestion
When you're developing and deploying software, one of the first things you need to figure out is how to go from the code you and your team write on your laptops to deploying your working application into a production environment where end users can interact with your software. Most development teams understand the importance of version control to coordinate code commits and build servers to compile and package their software, but continuous integration is such a big topic, we've decided to spend some time over the next few months going into detail with continuous integration and two of the most popular build servers, Jenkins Pipelines and GitHub Actions.
```

Author: @wordlee I believe we're talking about Jenkins, not just [Jenkins Pipeline](https://www.jenkins.io/doc/book/pipeline/) for this series. I asked about it in the channel (#content-initiative-2021-q4) so you can see the discussion there.

Author: @wordlee can I also suggest the following small edits to your proposed content?

When you're developing and deploying software, one of the first things to figure out is how to take your code and deploy your working application to a production environment where people can interact with your software.

Most development teams understand the importance of version control to coordinate code commits, and build servers to compile and package their software, but Continuous Integration (CI) is a big topic. Over the next few months, we’re going into detail about Continuous Integration and how two of the most popular build servers, Jenkins and GitHub Actions, can help with your CI processes.

## Why build servers are important

> Build servers have 3 main purposes:

Reviewer: I don't think we have this in the style guide, but I've always followed the Chicago manual of style for spelling out numbers below 100:

```suggestion
Build servers have three main purposes:
```

- Compiling committed code from your repository many times a day

> - Running thorough, automatic tests to validate code

Reviewer: I've removed thorough as that feels fairly subjective:

```suggestion
- Running automatic tests to validate code
```

- Creating deployable packages and handing off to your deployment tool

> Without a build server you're bogged down with manual processes and the needless time constraints caused by them. For example, without a build server:

Reviewer: Bogged is funny word :) feels very English and might be a struggle for ESL peeps. I also added complicated to the sentence as simplifying complex deployments is one of Octopus strongest points:

```suggestion
Without a build server you're slowed down by complicated, manual processes and the needless time constraints they introduce. For example, without a build server:
```

> - Your team will likely need to commit code before a daily deadline or during change windows.
> - Once that deadline passes, no one can commit again until someone manually creates and tests a build.
> - If there are problems with code, the deadlines and manual processes delay the fixes further.

Reviewer: Consistent punctuation:

```suggestion
- Your team will likely need to commit code before a daily deadline or during change windows
- After that deadline passes, no one can commit again until someone manually creates and tests a build
- If there are problems with code, the deadlines and manual processes further delay the fixes
```

> In this scenario, the team battles unnecessary hurdles. A build server, however, repeatedly does all this for you throughout the day, and without those human-caused delays.

Reviewer: ```suggestion
Without a build server, the team battles unnecessary hurdles that could be eliminated with automation. A build server, will repeatedly these tasks for you throughout the day, and without those human-caused delays.
```

But CI doesn’t just mean less time spent on manual tasks or the death of arbitrary deadlines, either. By automatically taking these steps many times a day, you fix problems sooner and your results become more predictable. Build servers ultimately help you deploy through your pipeline with more confidence.

> We’re taking deep dives into 2 CI platforms in the coming weeks, looking at how [Jenkins](https://www.jenkins.io/) or [GitHub Actions](https://github.com/features/actions) can help with your processes, plus how they complement Octopus.

Reviewer: Slight switch of focus:
```suggestion
We’re taking deep dives into two CI platforms in the coming weeks, looking at how [Jenkins](https://www.jenkins.io/) or [GitHub Actions](https://github.com/features/actions) can help with your processes, and how Octopus can help you get more out of your CI/CD pipeline.
```

> ## About Jenkins

Reviewer: From an SEO perspective, I understand what is X questions in titles really help.
```suggestion
## What is Jenkins
```

[Jenkins](https://www.jenkins.io/) is the most popular CI platform on the market. Open-source and free to use, you can run Jenkins standalone on most operating systems to automate the building and testing of your code.

One of Jenkins biggest benefits is its flexibility. It's a scalable platform, meaning you can expand its capabilities as your team or project needs grow. And thanks to its huge community, [there over 1800 plugins](https://plugins.jenkins.io/), making it easy to integrate with countless industry tools.
 
That means Jenkins is flexible enough to cover your CI needs and you can tailor it for other automation purposes too.

> ## About GitHub Actions

Reviewer: ```suggestion
## What is GitHub Actions
```
